### PR TITLE
Add `static_argnames` option to qjit

### PR DESCRIPTION
**Pull request by paul0403**
**Context:**
Adding a `static_argnames` option to qjit for users to configure static arguments by name.

**Description of the Change:**
Under the hood, this just maps the `static_argnames` to their argument indices and add to `static_argnums`.

**Benefits:**
Users can specify static arguments to jitted functions by name.

**Possible Drawbacks:**
Even more keyword arguments to qjit...

[sc-41335]

### CODE REVIEW

File: frontend/catalyst/jit.py [hunk 1]
```diff
@@ -123,6 +124,8 @@ def qjit(
             considered to be used by advanced users for low-level debugging purposes.
         static_argnums(int or Seqence[Int]): an index or a sequence of indices that specifies the
             positions of static arguments.
+        static_argnames(str or Seqence[str]): a string or a sequence of strings that specifies the
+            names of static arguments.
```
Comment (paul0403) [inline frontend/catalyst/jit.py:128]: changelog

**Comment by erick-xanadu** [comment]
I would replace the tests with lit tests that don't execute the function. They are quicker.

File: frontend/catalyst/jit.py [hunk 2]
```diff
@@ -430,6 +433,21 @@ def sum_abstracted(arr):
     if fn is None:
         return functools.partial(qjit, **kwargs)
 
+    # Map static_argnames to static_argnums
+    # no need to propagate static_argnames further
+    if static_argnames is not None:
```
Comment (rmoyard) [inline frontend/catalyst/jit.py:438]: What is Jax behaviour when you give an argname that is not part of the function?

**Review by rmoyard — COMMENTED** [review]
Looks good, just one question!

File: frontend/test/pytest/test_static_arguments.py [hunk 3]
```diff
@@ -221,6 +222,55 @@ def wrapper(x, c):
...
+                return x + y
+
+        assert q(1, 2) == 3
+
+        @qjit(static_argnames=("y"), static_argnums=[1])
+        def r(x, y):
+            if y < 10:
+                return x + y
+
+        assert r(1, 2) == 3
```
Comment (josh146) [inline frontend/test/pytest/test_static_arguments.py:272]: @paul0403 have you tested that it works with different combinations of defining and calling functions? E.g.,

- defining functions that use `*` and `/` in the signature
- defining functions that use default values in the signature
- calling with a mixture of keyword and positional values

File: frontend/catalyst/jit.py (hunk 2 above)
Comment (paul0403) [reply frontend/catalyst/jit.py:438]:
```python
@partial(jax.jit, static_argnames=('xx',))
def circuit(x,y):
    if x<10:
        return x+y

res = circuit(1, 2)
print(res)
```

```
ValueError: Jitted function has invalid argnames {'xx'} in static_argnames. Function does not take these args.
```

I will add this to qjit

**Comment by paul0403** [comment]
> I would replace the tests with lit tests that don't execute the function. They are quicker.

34a418c9b233c993ac5330261f70c4a3f2872d37

File: frontend/test/pytest/test_static_arguments.py (hunk 3 above)
Comment (paul0403) [reply frontend/test/pytest/test_static_arguments.py]: Through testing this I realized currently `static_argnums` do not support calls with default arguments (this is on main):
```python
@qjit(static_argnums=[1])
def f(y, x=9):
    if x < 10:
        return x + y
    return 42000

res = f(20)
print(res)
```

```
catalyst.utils.exceptions.CompileError: argnum 1 is beyond the valid range of [0, 1).
```

Note that jax works:
```python
@partial(jax.jit, static_argnums=[1])
def f(y, x=9):
    if x < 10:
        return x + y
    return 42000


res = f(20)
print(res)
```

```
29
```

I don't think this is an issue with `static_argname`. I propose we open a separate issue to fix it.

Comment (josh146) [reply frontend/test/pytest/test_static_arguments.py]: Nice, thanks @paul0403! My intuition was that this is worth testing as edge cases/bugs like this might appear :)

If this affects `static_argnum`, and would be resolved here once `static_argnum` is fixed, I am okay opening a separate issue to fix it.

File: frontend/catalyst/jit.py [hunk 4]
```diff
@@ -430,6 +433,33 @@ def sum_abstracted(arr):
...
```
Comment (dime10) [inline frontend/catalyst/jit.py:438]: I think my preference would be to handle this kwarg in the same place that we handle other options, so as not to have inconsistent handling of options by having the code in different places.

Comment (paul0403) [reply frontend/catalyst/jit.py:436]: The problem is, the jitted function is not available inside the `CompileOption` class, so this map won't be possible there.

Comment (paul0403) [reply frontend/catalyst/jit.py]: https://github.com/PennyLaneAI/catalyst/pull/1158/commits/bf05ee6eeaaaf9e9f8d81c75f5ac2f0f7b367851 is a workaround but I think this is more ugly

File: frontend/test/pytest/test_static_arguments.py (hunk 3 above)
Comment (paul0403) [reply frontend/test/pytest/test_static_arguments.py]: So I think none of the more complicated patterns work, but not because of `static_argname`, but because they never worked with `static_argnum` to begin with.

One particular case was handed out as an assessment: #1163 . In the issue I documented the failure. The apparent failure is in a verification, but I am not sure whether this means (a) the underlying mechanism is buggy and the verification failed as a by-product, or (b) the underlying mechanism is ok, but the verification itself is overly strict.

Comment (paul0403) [reply frontend/test/pytest/test_static_arguments.py]: @josh146 I suggest we finish the simple base case of `static_argnames` for release, since the other failed cases were already there with `static_argnums` as well.

**Comment by paul0403** [comment]
> Another integration test we might want to consider is with decorators (like vmap, qnode, grad, etc), since this functionality inspects the function signature, there is a chance decorators might not propagate this info properly.

Really conveniently, `inspect` module has a `inspect.signature()`, which can take in any callable, not just plain functions.
https://github.com/PennyLaneAI/catalyst/pull/1158/commits/7e80645356ea314c4374d9e48f95f6581f76a7e6

File: frontend/test/pytest/test_static_arguments.py (hunk 3 above)
Comment (josh146) [reply frontend/test/pytest/test_static_arguments.py]: that works for me!

File: frontend/catalyst/tracing/type_signatures.py [hunk 5]
```diff
@@ -156,6 +156,34 @@ def split_static_args(args, static_argnums):
...
+Function does not take these args.
+                """
+            )
+        new_static_argnums.append(fn_argnames.index(static_argname))
+
+    # Remove potential duplicates from static_argnums and static_argnames
+    new_static_argnums = list(dict.fromkeys(new_static_argnums))
+    new_static_argnums.sort()
+
+    return tuple(new_static_argnums)
```
Comment (joeycarter) [inline frontend/catalyst/tracing/type_signatures.py:184]:
```suggestion
    return tuple(sorted(set(new_static_argnums)))
```

Would this work too? Simpler, more expressive and has the same functionality.

File: frontend/test/pytest/test_static_arguments.py [hunk 6]
```diff
@@ -221,6 +221,75 @@ def wrapper(x, c):
...
+        # pylint: disable=unused-argument, function-redefined
+        """Test static arguments specified by names"""
+
+        @qjit(static_argnames="y")
+        def f(x, y):
+            return
+
+        assert set(f.compile_options.static_argnums) == {1}
+
+        with pytest.raises(ValueError, match="qjitted function has invalid argname {yy}"):
```
Comment (joeycarter) [inline frontend/test/pytest/test_static_arguments.py:234]:
```suggestion
        with pytest.raises(ValueError, match="The `static_argnames` argument 'yy' in `qjit` is not a valid argument"):
```

If you plan to change the error message.

File: frontend/test/pytest/test_static_arguments.py [hunk 7]
```diff
@@ -221,6 +221,75 @@ def wrapper(x, c):
...
+
+            @qjit(static_argnames="yy")
+            def f_badname(x, y):
+                return
+
+        with pytest.raises(ValueError, match="qjitted function has invalid argname {yy}"):
```
Comment (joeycarter) [inline frontend/test/pytest/test_static_arguments.py:240]:
```suggestion
        with pytest.raises(ValueError, match="The `static_argnames` argument 'yy' in `qjit` is not a valid argument"):
```

If you plan to change the error message.

File: frontend/catalyst/tracing/type_signatures.py [hunk 8]
```diff
@@ -156,6 +156,34 @@ def split_static_args(args, static_argnums):
...
+        static_argnames = [static_argnames]
+
+    for static_argname in static_argnames:
+        if static_argname not in fn_argnames:
+            raise ValueError(
+                f"""
+qjitted function has invalid argname {{{static_argname}}} in static_argnames.
+Function does not take these args.
+                """
+            )
```
[the remark this answers is not in the document]
Comment (paul0403) [reply frontend/catalyst/tracing/type_signatures.py:177]: We track the same error message as jax https://github.com/jax-ml/jax/blob/2b7b0742a426565632a627699fe3cb48e4f17b89/jax/_src/api_util.py#L217

File: frontend/catalyst/jit.py (hunk 4 above)
Comment (paul0403) [reply frontend/catalyst/jit.py]: I am now merging static_argname into static_argnum in the QJIT class's __init__.

File: frontend/catalyst/tracing/type_signatures.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (joeycarter) [reply frontend/catalyst/tracing/type_signatures.py:177]:
```suggestion
            raise ValueError(
                f"""
qjitted function has invalid argname {static_argname} in static_argnames.
Function does not take these args.
                """
            )
```

Ah, in that case, maybe just remove the extra curly braces around `{static_argname}` so it's the same as in JAX.

File: doc/releases/changelog-dev.md [hunk 9]
```diff
@@ -185,6 +185,30 @@
...
+    if x < 10 and y < 10:
+        return x + y
+
+  res_f = f(1, 2)
+  res_g = g(3, 4)
+  print(res_f, res_g)
+  ```
+
+  ```pycon
+  >>> 3 7
```
[the remark this answers is not in the document]
Comment (paul0403) [reply doc/releases/changelog-dev.md:209]: done

File: frontend/catalyst/tracing/type_signatures.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (paul0403) [reply frontend/catalyst/tracing/type_signatures.py:177]: Jax error message actually contains the braces:
```python
@partial(jax.jit, static_argnames=["xx"])
def f(y, x):
    if x < 10:
        return x + y
    return 42000


res = f(20, 3)
print(res)
```
ValueError: Jitted function has invalid argnames {'xx'} in static_argnames. Function does not take these args.
```

```

[the remark this answers is not in the document]
Comment (joeycarter) [reply frontend/catalyst/tracing/type_signatures.py:177]: Discussed offline, putting here for reference. If we have multiple invalid `static_argnames`, and want to print the same error message as JAX, we should loop over all static argnames, collect the invalid ones, and finally raise the exception. For example, in JAX:

```python
from functools import partial
import jax

@partial(jax.jit, static_argnames=["xx", "yy"])
def f(y, x):
    if x < 10:
        return x + y
    return 42000


res = f(20, 3)
print(res)
```

```
ValueError: Jitted function has invalid argnames {'xx', 'yy'} in static_argnames. Function does not take these args.
```

[the remark this answers is not in the document]
Comment (joeycarter) [reply frontend/catalyst/tracing/type_signatures.py:177]: If we do it this way we should also add a test case for it :slightly_smiling_face:

File: frontend/catalyst/tracing/type_signatures.py [hunk 10]
```diff
@@ -156,6 +156,38 @@ def split_static_args(args, static_argnums):
...
+            continue
+        non_existent_args.append(static_argname)
+
+    if non_existent_args:
+        raise ValueError(
+            f"""
+qjitted function has invalid argname {set(non_existent_args)} in static_argnames.
+Function does not take these args.
+            """
+        )
```
Comment (joeycarter) [inline frontend/catalyst/tracing/type_signatures.py:183]: Last nitpick I promise :slightly_smiling_face:

JAX doesn't put the new line before "Function does not take these args." Could we do this instead?

```suggestion
        raise ValueError(
            f"qjitted function has invalid argname {set(non_existent_args)} in static_argnames. "
            "Function does not take these args."
        )
```

Comment (paul0403) [reply frontend/catalyst/tracing/type_signatures.py]: done

File: frontend/catalyst/tracing/type_signatures.py [hunk 11]
```diff
@@ -156,6 +156,41 @@ def split_static_args(args, static_argnums):
...
+        if static_argname in fn_argnames:
+            new_static_argnums.append(fn_argnames.index(static_argname))
+            continue
+        non_existent_args.append(static_argname)
+
+    if non_existent_args:
+        non_existent_args_str = "{"
+        for arg in non_existent_args:
+            non_existent_args_str += "'" + arg + "', "
+        non_existent_args_str = non_existent_args_str[:-2] + "}"
```
Comment (joeycarter) [inline frontend/catalyst/tracing/type_signatures.py:181]: Just so I understand, is the purpose of this loop over `non_existent_args` just to preserve the order of arguments as they appear in `static_argnames`, since `set(non_existent_args)` does not preserve that order?

Comment (paul0403) [reply frontend/catalyst/tracing/type_signatures.py:181]: Yes; I tried various things but none seemed to work, so I have to resort to the raw way of doing it

Comment (paul0403) [reply frontend/catalyst/tracing/type_signatures.py:181]: `set(sorted(set(non_existent_args)))` also does not preserve the order

Comment (paul0403) [reply frontend/catalyst/tracing/type_signatures.py:181]: I do think `sorted(set(non_existent_args))` will preserve the order, but the outmost set becomes unordered again (since it's a set).

Really, we just want the braces in the error msg, so we can just build the string ourselves if making sets happy is too difficult.

Comment (joeycarter) [reply frontend/catalyst/tracing/type_signatures.py:181]: Yeah I think this is fine. Here's a one-liner that essentially does the same thing. Feel free to use it if you like.

```python
'{' + ', '.join(repr(item) for item in non_existent_args) + '}'
```

For example:

```pycon
>>> non_existent_args = ['b', 'a', 'c', '[d]', 'e']
>>> result = '{' + ', '.join(repr(item) for item in non_existent_args) + '}'
>>> print(result)
{'b', 'a', 'c', '[d]', 'e'}
```

Comment (paul0403) [reply frontend/catalyst/tracing/type_signatures.py]: https://github.com/PennyLaneAI/catalyst/pull/1158/commits/771f2fbd9a0608a7779159a636c096a05f78a015

Comment (joeycarter) [reply frontend/catalyst/tracing/type_signatures.py]: [Don't fear the `repr`!](https://en.wikipedia.org/wiki/(Don%27t_Fear)_The_Reaper)

**Review by joeycarter — APPROVED** [review]

File: frontend/test/pytest/test_static_arguments.py (hunk 3 above)
Comment (josh146) [reply frontend/test/pytest/test_static_arguments.py]: @paul0403 if you haven't already, would you be able to open an issue detailing the more advanced cases that don't work with `static_argnums`? https://github.com/PennyLaneAI/catalyst/pull/1158#discussion_r1781382066

Comment (paul0403) [reply frontend/test/pytest/test_static_arguments.py]: > @paul0403 if you haven't already, would you be able to open an issue detailing the more advanced cases that don't work with `static_argnums`? [#1158 (comment)](https://github.com/PennyLaneAI/catalyst/pull/1158#discussion_r1781382066)

Already did, and actually we sent this out as an assessment! #1163

Comment (paul0403) [reply frontend/test/pytest/test_static_arguments.py]: (As far as I can tell, the default value issue seems like the root cause of all other patterns' failures.)
